zerver/tests/test_signup.py

```diff
@@ -3284,6 +3285,42 @@ def test_create_realm_ldap_email(self) -> None:
                 "newuser_email@zulip.com", self.ldap_password("newuser_with_email")
             )
 
+    @override_settings(OPEN_REALM_CREATION=True)
+    def test_create_realm_race(self) -> None:
+        email = "romeo@zulip.com"
+        password = "test"
+
+        with patch("zerver.views.registration.do_create_realm", side_effect=IntegrityError):
+            internal_realm = get_realm(settings.SYSTEM_BOT_REALM)
+            notification_bot = get_system_bot(settings.NOTIFICATION_BOT, internal_realm.id)
+            signups_stream, _ = create_stream_if_needed(notification_bot.realm, "signups")
+
+            string_id = "zulipracetest"
+            # Make sure the realm does not exist
+            with self.assertRaises(Realm.DoesNotExist):
+                get_realm(string_id)
+
+            # Create new realm with the email
+            result = self.client_post("/new/", {"email": email})
+            self.assertEqual(result.status_code, 302)
+            self.assertTrue(result["Location"].endswith(f"/accounts/new/send_confirm/{email}"))
+            result = self.client_get(result["Location"])
+            self.assert_in_response("Check your email so we can get started.", result)
+
+            # Check confirmation email has the correct subject and body, extract
+            # confirmation link and visit it
+            confirmation_url = self.get_confirmation_url_from_outbox(
+                email,
+                email_subject_contains="Create your Zulip organization",
+                email_body_contains="You have requested a new Zulip organization",
+            )
+            result = self.client_get(confirmation_url)
+            self.assertEqual(result.status_code, 200)
```

Comment:
Is it expected that this step will have created a realm? I think it should have just shown a registration page, in which case possibly much of this block may not actually be necessary for the test.

Reply:
I started with the test that proved realm creation, and then backed it out to where the error occurred on exception. The process isn't super clear what is required for bootstrap and not. I'll play with culling more of it.